Beautify and better examples

andria-dev opened the pull request:
The JSX example was poor due to its misuse of `do`. `do`, in this example, was being used where a ternary is often used, so to ensure the reason to use `do` was more effectively communicated I designed a more complicated example where a ternary would be unpreferable. Semicolons were being used off and on, therefore, I removed them as to avoid confusion. Increased readability in the conditional section and removed unnecessary curly braces.

Commit:
Beautify and better examples

The JSX example was poor due to its misuse of `do`. `do`, in this example, was being used where a ternary is often used, so to ensure the reason to use `do` was more effectively communicated I designed a more complicated example where a ternary would be unpreferable. Semicolons were being used off and on, therefore, I removed them as to avoid confusion. Increased readability in the conditional section and removed unnecessary curly braces.

pitaj commented on README.md:
```diff
@@ -38,10 +41,17 @@ return (
+        if (!loggedIn) {
           <LoginButton />
+        } else {
+          if (membershipStatus === 'basic') {
```
This can be an `else if` instead of embedding all of them.

pitaj commented on README.md:
```diff
@@ -38,10 +41,17 @@ return (
+          } else if (membershipStatus === 'premium') {
+            <PremiumBadge />
+          } else {
+            alertUserOfPaymentIssues()
```
It isn't recommended to do side-effects in a render function

pitaj commented on README.md:
```diff
@@ -24,9 +24,12 @@ Use conditional statements as expressions, instead of awkward nested ternaries:
-  if (foo()) { f() }
-  else if (bar()) { g() }
-  else { h() }
+  if (foo())
```
It's a very standard practice in JS (and other languages with optional curly-braces) to just _always_ use them.

ChrisBrownie55 replied:
The curly braces aren't as much the problem as much as putting both the condition and the statement on the same line, separating them was just to improve readability.

Commit:
Made fixes according to comments

Commit:
Change `tmp * tmp` to `tmp ** 2`

It's the same thing just being less redundant. Like, temp squared plus one, instead of temp times time plus one.

Commit:
Change first example to be more realistic

The use case needs to include the use of `f()` twice as a stored variable (`tmp`) but squaring can be done without using `tmp` twice, so instead `tmp` to the power of `tmp` plus one is now a better example of when you would actually need a `do` statement.

Commit:
Unnecessary semicolon removal

The first semicolon was fine but as most of the other semicolons have been removed I find no need to keep that one here, but the second semicolon would have actually caused an error as per one of the current pull requests, so it was easier to remove it to avoid more confusion.

ChrisBrownie55 commented:
Any word on this being able to be merged?

ChrisBrownie55 commented:
Ping?

jonathantneal commented:
These really are great examples. Thanks, @ChrisBrownie55. I hope these are merged.

ChrisBrownie55 commented:
@jonathantneal Thanks, it's been a hot minute since I introduced this so I'm thinking it won't get merged though lol

Akxe commented:
I would suggest adding null coalescing example:

```ts
declare const cache: Map<T, R>;

function heavyTransform(obj: T) {
  return cache.get(obj) ?? do {
    const computedValue = ...;
    cache.set(obj, computedValue);
    computedValue 
  }
}
```

Or similar... I always dream of this when writing this kind of look ups...

00ff0000red commented:
One of those examples aren't that strong, the if else could easily be a ternary:
```js
const x = (
  foo()
    ? a
    : bar()
      ? b
      : c
)();
```

I'd suggest the following:
* use more `const` (really important part of FP and value-returning blocks in general is immutability)
* give an example where someone would use a `let` and seperate assignments
* give an example that requires more that just a value in the if statement body (@Akxe's example almost fits this)

For the second, a try/catch would be a great example, as a ternary wouldn't fit at all.
```js
let result;
try {
    const random = Math.random();
    const powed = random ** random;
    result = calc_something(powed, random);
} catch {
    // aww, `calc_something` threw, fallback
    result = calc_something_else(0.5, 4);
}
```

for (3), something like this might be better:
```js
const x = do {
  if ( foo() ) {
    a();
  } else if ( bar() ) {
    const random = Math.random();
    const temp = get_nth(random ** random);
    temp;
  } else {
    c();
  }
};
```
as a ternary would be far too verbose, and it would require a iife to match the second if.

*^ these are 2 poor examples, please do better than I did ;)*

ljharb commented:
Nested ternaries are often considered poor style, and common styleguides/eslint configs forbid them.

rkirsling commented:
Obligatory note that those aren't nested; they are chained.

ljharb commented:
@rkirsling they're definitely nested; if you put parens to match default precedence, one of the ternaries will be contained within the truthy or falsy branch of the other ternary.

rkirsling commented:
It's a critical distinction because we don't call `if {} else if {} else {}` "nested `if`s", even though `else if` is likewise not its own thing under the hood.

Recommending against _nesting_ is pretty uncontroversial, but `a ? b : c ? d : e` is a baby too often thrown out with that bathwater.

If there's one hill I'll die on in programming language design, it's definitely this. :smile:

00ff0000red commented:
Personally, I see things exactly the same way that @ljharb sees it, I consider `? : ?  :` to be "nested," even though they don't have the parenthesis like `? : (?  :)` does, and am personally okay with writing `if () { } else { if () {} else {} }`, so I am likely to be a minority here regarding sytlistic preferences.

Now, even though I'm one of the (*apparently few*) people who are perfectly fine with reading and using ternaries, I still see value in block scopes that can return values, but attempting to create one thing in a language to obsolete another one solely due to syntax seems like a very weak argument. That is why I strongly suggest showing where "do blocks" cannot easily be replaced with another syntatical construct.

ChrisBrownie55 commented:
I think this PR is dead. However, for the sake of readability, multiple ternary chains can become less readable in some cases but that doesn't mean you can't use them if they suit your need. If this was all about the React example, I added more else-if-statements because, from my experience, instead of chaining ternaries together (especially when the values or conditions are multiple lines long), it's often easier to extract the logic out of the return-statement.
